### PR TITLE
Refactoring sharding package for proposer/client separation

### PR DESCRIPTION
**Pull request by prestonvanloon**
This builds on the work of @nisdas in #82.

### Why refactor?
We started this package under the original sharding design that would have only a collator. Now that we need to split responsibilities into 2+ roles, it makes sense to keep things isolated in their own packages. 

Following @nisdas's [design](https://docs.google.com/document/d/1__74pTWFpzfLLJ8017XeWQWJfs89YgTGxuefjkFUAaY/edit) along with suggestions in: 
- [Effective go](https://golang.org/doc/effective_go.html)
- [CodeReviewComments](https://github.com/golang/go/wiki/CodeReviewComments#interfaces)

We split up the _client_ aspect of the roles into a client package. This package is responsible for the actual communications (i.e. connecting to a running geth node, initializing the SMC, handling account unlocking). Then we can have a _collator_ that is more focused in its own package with responsibilities such as listening for block headers, joining the collator pool, etc. Furthermore, we start the _proposer_ package to run with its own responsibilities. 

### Why have `collator_client.go` and `collator.go`? 
Good question! At first, I didn't like this, but the `collator_client` could/should expose the public methods that the `geth/shardingcmd` can call (i.e. `Start` and `NewCollator`) while the interesting code lives in `collator.go`. This felt helpful to separate for testing purposes, but I can be convinced out of this pattern if your argument is strong enough :) 

This supersedes (and includes) @nisdas's work in #82.
Closes #82 
Closes #83 
Required for #79 and #80 

TODO (in future PR):
- More testing of `collator`
- Move common testing methods (mock client) into `sharding/client/testing`

### CODE REVIEW

**Comment by terenc3t** [comment]
Looks good! Two questions I'd like to discuss:

1. The necessity to separate proposer and collator into different packages. It means whenever a new actor is introduced, it needs to be its own package. What benefit do we get out of that
2. I don't think `initSMC` should be on the client level because not all types of client will need to interact with SMC

**Comment by prestonvanloon** [comment]
Good questions!

1. The benefit is a clear separation of concerns as new responsibilities join sharding.This seems like it would scale better than 1 sharding package with many files in it. Is there a reason we would prefer to have all of them in one package? 
2. Hmm yes, perhaps the `smc.go` could live in contracts? Would the executor not need access to the SMC?

**Comment by prestonvanloon** [comment]
Just to be clear, I am OK with one sharding package. 
I'd like to reference a definitive authority on which is preferred, if one exists.

File: sharding/client/smc.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/sharding"
+	"github.com/ethereum/go-ethereum/sharding/contracts"
+)
+
+// initSMC initializes the sharding manager contract bindings.
+// If the SMC does not exist, it will be deployed.
+func initSMC(c *shardingClient) (*contracts.SMC, error) {
```
Comment (rauljordan) [inline sharding/client/smc.go]: Can we put this function in the client.go file? The client is the one that inits the SMC and it doesn't make sense for this file to exist at all if we only envision having a few functions or just 1 in it.

File: sharding/client/client.go [hunk 2]
```diff
@@ -1,6 +1,4 @@
-package sharding
-
-//go:generate abigen --sol contracts/sharding_manager.sol --pkg contracts --out contracts/sharding_manager.go
+package client
```
Comment (rauljordan) [inline sharding/client/client.go]: Let's please rename this to sharding_client.go. client is ambiguous with the ethclient that is a part of the struct.

Comment (prestonvanloon) [reply sharding/client/client.go]: This isn't related to the filename, but the package path. Do we need the redundancy of sharding/sharding_client/client.go?

File: sharding/client/smc.go (hunk 1 above)
Comment (prestonvanloon) [reply sharding/client/smc.go]: Sure. I don't feel very strongly about having this method in its own file so I'll move it.

File: sharding/client/client.go [hunk 3]
```diff
@@ -30,17 +27,28 @@ const (
...
 )
 
-// Client for Collator. Communicates to Geth node via JSON RPC.
-type collatorClient struct {
-	endpoint string             // Endpoint to JSON RPC
-	client   *ethclient.Client  // Ethereum RPC client.
-	keystore *keystore.KeyStore // Keystore containing the single signer
-	ctx      *cli.Context       // Command line context
-	smc      *contracts.SMC     // The deployed sharding management contract
+// General Client for Collator. Communicates to Geth node via JSON RPC.
```
Comment (terenc3t) [inline sharding/client/client.go]: This is not only for collator, it's for proposer and future actors too

File: sharding/client/client.go [hunk 4]
```diff
@@ -93,42 +100,22 @@ func (c *collatorClient) Start() error {
...
-	}
-
-	// Deposit 100ETH into the collator set in the SMC. Checks if account
-	// is already a collator in the SMC (in the case the client restarted).
-	// Once that's done we can subscribe to block headers.
-	//
-	// TODO: this function should store the collator's SMC index as a property
-	// in the client's struct
-	if err := joinCollatorPool(c); err != nil {
+	smc, err := initSMC(c)
```
Comment (terenc3t) [inline sharding/client/client.go]: I don't know if executor node uses SMC, we might have to change this for phase 3

File: sharding/collator/smc.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/sharding"
+	"github.com/ethereum/go-ethereum/sharding/client"
+)
+
+// joinCollatorPool checks if the account is a collator in the SMC. If
+// the account is not in the set, it will deposit 100ETH into contract.
+func joinCollatorPool(c client.Client) error {
```
Comment (terenc3t) [inline sharding/collator/smc.go]: This is a Collator specific function. Should we move it under `collator.go` and delete `smc.go` completely?

File: sharding/client/client.go (hunk 4 above)
Comment (prestonvanloon) [reply sharding/client/client.go]: The executor node won't use this client anyway, it doesn't need main chain communication, right?

File: sharding/collator/smc.go (hunk 5 above)
Comment (prestonvanloon) [reply sharding/collator/smc.go]: Sure, sounds good to me

File: sharding/client/client.go (hunk 4 above)
Comment (terenc3t) [reply sharding/client/client.go]: That's true

File: sharding/client/smc.go (hunk 1 above)
Comment (prestonvanloon) [reply sharding/client/smc.go]: Done

File: sharding/client/client.go (hunk 2 above)
Comment (prestonvanloon) [reply sharding/client/client.go]: Let's revisit this in the future. I'm going to leave for now.

File: sharding/client/client.go (hunk 3 above)
Comment (prestonvanloon) [reply sharding/client/client.go]: Done

**Comment by prestonvanloon** [comment]
Travis failed some times, but tests pass so i'll merge anyway

**Pull request merged**